examples/nn.dx

```diff
@@ -0,0 +1,246 @@
+' # Neural Networks
+
+include "plot.dx"
+
+' ## NN Prelude
+
+
+def relu (input : Float) : Float =
+  select (input > 0.0) input 0.0
+
+instance [Add a, Add b] Add (a & b)
+  add = \(a, b) (c, d). ( (a + c), (b + d))
+  sub = \(a, b) (c, d). ( (a - c), (b - d))
+  zero = (zero, zero)
+
+instance [VSpace a, VSpace b] VSpace (a & b)
+  scaleVec = \ s (a, b) . (scaleVec s a, scaleVec s b)
+
+def Layer (inp:Type) (out:Type) (param:Type) : Type=
+  {forward: param -> inp -> out &
+   init: Key -> param
+  }
+
+def forward (l:Layer i o p) (p : p) (x : i): o =
+  (getAt #forward l) p x
+
+def init (l:Layer i o p) (k:Key)  : p  =
+  (getAt #init l) k
+
+
+' ## Layers 
+
+' Dense layer
+
+def Dense (a:Type) (b:Type) : Type =
+  Layer (a=>Float) (b=>Float) ((a=>b=>Float) & (b=>Float))
```

Comment:
Starting a new thread here to avoid talking about too many things in the other one.

> But morally is it bad to have an alias? It makes the type clearer, and it is helpful in debugging since you can specify the type with less keystroke. (For instance below in the lenet you could type the modules.)

My personal opinion is that the name of the alias isn't "correct" from a perspective of the meaning of the aliased type. A value of type `Dense a b` is (currently) an implementation of  an arbitrary neural network that happens to follow the same signature as `dense`. So for instance you could write
```
confusing : Dense a b = {
  forward = (\weight, bias) x. zero)
  init = arb
}
```
But that feels weird to me, and I don't know why you'd want to say that that "is a Dense" in any meaningful sense?

Personally, something like
```
def DenseParams (a:Type) (b:Type) : Type = ((a=>b=>Float) & (b=>Float))
def dense (a:Type) (b:Type) : Layer (a=>Float) (b=>Float) (DenseParams a b) = ...
```
better matches my intuition about what `dense` is: it's a layer that takes `(a=>Float)` to `(b=>Float)` and is parameterized by a particular set of parameters.

Does that make sense? I'm not sure I'm explaining my thoughts clearly here.

Reply:
Okay, pushed a new version with Param aliases, but an explicit data / newtype for Layer. 

The main purpose of the data here is to hide the type details of the layer implementation. 

You still get network types like this that expose all the internal parameters, but I guess with records that is not so bad. Also this will at least let us write `Show` and `Arb` and maybe things like `Read` for layers. 

```haskell
:t simple
((h1:Type)
 -> Layer
      ((Fin 2) => Float32)
      ((Fin 2) => Float32)
      ( (((Fin 2) => h1 => Float32) & (h1 => Float32))
      & ((h1 => (Fin 2) => Float32) & ((Fin 2) => Float32))))
```

Reply:
Just wanted to say: god that's nice